File: app/src/main/java/cn/ifafu/ifafu/data/local/RepositoryImpl.java
package cn.ifafu.ifafu.data.local;

import androidx.annotation.Nullable;

import org.greenrobot.greendao.AbstractDao;

import java.util.List;

import cn.ifafu.ifafu.app.Constant;
import cn.ifafu.ifafu.dao.CourseDao;
import cn.ifafu.ifafu.dao.DaoSession;
import cn.ifafu.ifafu.dao.ElecCookieDao;
import cn.ifafu.ifafu.dao.ElecQueryDao;
import cn.ifafu.ifafu.dao.ElecUserDao;
import cn.ifafu.ifafu.dao.ExamDao;
import cn.ifafu.ifafu.dao.ScoreDao;
import cn.ifafu.ifafu.dao.SyllabusSettingDao;
import cn.ifafu.ifafu.dao.TokenDao;
import cn.ifafu.ifafu.dao.UserDao;
import cn.ifafu.ifafu.data.entity.Course;
import cn.ifafu.ifafu.data.entity.ElecCookie;
import cn.ifafu.ifafu.data.entity.ElecQuery;
import cn.ifafu.ifafu.data.entity.ElecUser;
import cn.ifafu.ifafu.data.entity.Exam;
import cn.ifafu.ifafu.data.entity.Score;
import cn.ifafu.ifafu.data.entity.SyllabusSetting;
import cn.ifafu.ifafu.data.entity.Token;
import cn.ifafu.ifafu.data.entity.User;
import cn.ifafu.ifafu.util.SPUtils;

public class RepositoryImpl implements Repository {

    private UserDao userDao;
    private TokenDao tokenDao;
    private CourseDao courseDao;
    private ScoreDao scoreDao;
    private ExamDao examDao;
    private SyllabusSettingDao syllabusSettingDao;
    private ElecQueryDao elecQueryDao;
    private ElecCookieDao elecCookieDao;
    private ElecUserDao elecUserDao;

    private User user;

    private static Repository INSTANCE;

    private RepositoryImpl() {
        DaoSession daoSession = DaoManager.getInstance().getDaoSession();
        userDao = daoSession.getUserDao();
        tokenDao = daoSession.getTokenDao();
        courseDao = daoSession.getCourseDao();
        scoreDao = daoSession.getScoreDao();
        examDao = daoSession.getExamDao();
        syllabusSettingDao = daoSession.getSyllabusSettingDao();
        elecQueryDao = daoSession.getElecQueryDao();
        elecCookieDao = daoSession.getElecCookieDao();
        elecUserDao = daoSession.getElecUserDao();
    }

    public static Repository getInstance() {
        if (INSTANCE == null) {
            synchronized (RepositoryImpl.class) {
                if (INSTANCE == null) {
                    synchronized (RepositoryImpl.class) {
                        INSTANCE = new RepositoryImpl();
                    }
                }
            }
        }
        return INSTANCE;
    }

    @Override
    public User getLoginUser() {
        if (user == null) {
            synchronized (RepositoryImpl.class) {
                if (user == null) {
                    String account = SPUtils.get(Constant.SP_USER_INFO).getString("account");
                    user = userDao.load(account);
                }
            }
        }
        return user;
    }

    @Override
    public List<User> getAllUser() {
        return userDao.loadAll();
    }

    @Override
    public void saveUser(User user) {
        this.user = user;
        SPUtils.get(Constant.SP_USER_INFO).putString("account", user.getAccount());
        userDao.insertOrReplace(user);
    }

    @Override
    public void saveLoginUser(User user) {
        this.user = user;
        SPUtils.get(Constant.SP_USER_INFO).putString("account", user.getAccount());
    }

    @Override
    public void deleteUser(User user) {
        if (user.getAccount().equals(getLoginUser().getAccount())) {
            this.user = null;
        }
        userDao.deleteByKey(user.getAccount());
    }

    @Override
    public List<Course> getAllCourses() {
        return courseDao.queryBuilder()
                .where(CourseDao.Properties.Account.eq(getLoginUser().getAccount()))
                .list();
    }

    @Override
    public List<Course> getCourses(boolean local) {
        return courseDao.queryBuilder()
                .where(CourseDao.Properties.Account.eq(getLoginUser().getAccount()),
                        CourseDao.Properties.Local.eq(local))
                .list();
    }

    @Nullable
    @Override
    public Course getCourseById(long id) {
        return courseDao.load(id);
    }

    @Override
    public void saveCourse(Course course) {
        courseDao.insertOrReplace(course);
    }

    @Override
    public void saveCourse(List<Course> courses) {
        courseDao.insertOrReplaceInTx(courses);
    }

    @Override
    public void deleteCourse(List<Course> courses) {
        courseDao.deleteInTx(courses);
    }

    @Override
    public void deleteCourse(Course course) {
        courseDao.delete(course);
    }

    @Override
    public void deleteAllOnlineCourse() {
        deleteCourse(getCourses(false));
    }

    @Override
    public SyllabusSetting getSyllabusSetting() {
        return syllabusSettingDao.load(getLoginUser().getAccount());
    }

    @Override
    public void saveSyllabusSetting(SyllabusSetting syllabusSetting) {
        syllabusSettingDao.insertOrReplace(syllabusSetting);
    }

    @Override
    public List<Score> getAllScores() {
        return scoreDao.queryBuilder()
                .where(ScoreDao.Properties.Account.eq(getLoginUser().getAccount()))
                .list();
    }

    @Override
    public List<Score> getScoresByYear(String year) {
        return scoreDao.queryBuilder()
                .where(ScoreDao.Properties.Account.eq(getLoginUser().getAccount()),
                        ScoreDao.Properties.Year.eq(year))
                .list();
    }

    @Override
    public List<Score> getScoresByTerm(String term) {
        return scoreDao.queryBuilder()
                .where(ScoreDao.Properties.Account.eq(getLoginUser().getAccount()),
                        ScoreDao.Properties.Term.eq(term))
                .list();
    }

    @Override
    public Score getScoreById(long id) {
        return scoreDao.load(id);
    }

    @Override
    public List<Score> getScores(String year, String term) {
        return scoreDao.queryBuilder()
                .where(ScoreDao.Properties.Account.eq(getLoginUser().getAccount()),
                        ScoreDao.Properties.Year.eq(year),
                        ScoreDao.Properties.Term.eq(term))
                .list();
    }

    @Override
    public void deleteScore(String year, String term) {
        scoreDao.deleteInTx(getScores(year, term));
    }

    @Override
    public void deleteScore(List<Score> scores) {
        scoreDao.deleteInTx(scores);
    }

    @Override
    public void deleteAllScore() {
        scoreDao.deleteInTx(getAllScores());
    }

    @Override
    public void saveScore(List<Score> scores) {
        scoreDao.insertOrReplaceInTx(scores);
    }

    @Override
    public List<Exam> getAllExams() {
        return examDao.queryBuilder()
                .where(ExamDao.Properties.Account.eq(getLoginUser().getAccount()))
                .list();
    }

    @Override
    public List<Exam> getExams(String year, String term) {
        return examDao.queryBuilder()
                .where(ExamDao.Properties.Account.eq(getLoginUser().getAccount()),
                        ExamDao.Properties.Year.eq(year),
                        ExamDao.Properties.Term.eq(term))
                .list();
    }

    @Override
    public void saveExam(List<Exam> exams) {
        examDao.insertOrReplaceInTx(exams);
    }

    @Override
    public Token getToken(String account) {
        return tokenDao.load(account);
    }

    @Override
    public void saveToken(Token token) {
        tokenDao.insertOrReplace(token);
    }

    @Override
    public void clearAllData() {
        for (AbstractDao<?, ?> allDao : DaoManager.getInstance().getDaoSession().getAllDaos()) {
            allDao.deleteAll();
        }
    }

    @Override
    public ElecQuery getElecQuery() {
        return elecQueryDao.load(getLoginUser().getAccount());
    }

    @Override
    public void saveElecQuery(ElecQuery elecQuery) {
        elecQueryDao.insertOrReplace(elecQuery);
    }

    @Override
    public ElecCookie getElecCookie() {
        return elecCookieDao.load(getLoginUser().getAccount());
    }

    @Override
    public void saveElecCookie(ElecCookie cookie) {
        elecCookieDao.insertOrReplace(cookie);
    }

    @Override
    public ElecUser getElecUser() {
        return elecUserDao.load(getLoginUser().getAccount());
    }

    @Override
    public void saveElecUser(ElecUser elecUser) {
        elecUserDao.insertOrReplace(elecUser);
    }
}

File: app/src/main/java/cn/ifafu/ifafu/data/RepositoryImpl.kt
package cn.ifafu.ifafu.data

import android.content.Context
import androidx.room.Room
import cn.ifafu.ifafu.app.Constant
import cn.ifafu.ifafu.data.local.AppDatabase
import cn.ifafu.ifafu.entity.*
import cn.ifafu.ifafu.util.SPUtils
import java.util.*

object RepositoryImpl : Repository {

    private lateinit var db: AppDatabase

    fun init(context: Context) {
        db = Room.databaseBuilder(context, AppDatabase::class.java, "ifafu")
                .fallbackToDestructiveMigration()
                .build()
    }


    var account: String = ""
        get() {
            return field.ifEmpty { SPUtils.get(Constant.SP_USER_INFO).getString("account") }
        }
        private set

    override fun getInUseUser(): User? {
        return db.userDao.user(account)
    }

    override fun getAllUsers(): List<User> {
        return db.userDao.allUser()
    }

    override fun saveUser(user: User) {
        account = user.account
        SPUtils.get(Constant.SP_USER_INFO).putString("account", user.account)
        db.userDao.save(user)
    }

    override fun saveLoginUser(user: User) {
        account = user.account
        SPUtils.get(Constant.SP_USER_INFO).putString("account", user.account)
    }

    override fun deleteUser(user: User) {
        db.userDao.delete(user)
    }

    override fun getAllCourses(): List<Course> {
        return db.courseDao.allCourses(account)
    }

    override fun getCourses(local: Boolean): List<Course> {
        return db.courseDao.allCourses(account, local)
    }

    override fun getCourseById(id: Long): Course? {
        return db.courseDao.course(id)
    }

    override fun saveCourse(course: Course) {
        db.courseDao.save(course)
    }

    override fun saveCourse(courses: List<Course>) {
        db.courseDao.save(*courses.toTypedArray())
    }

    override fun deleteCourse(courses: List<Course>) {
        db.courseDao.delete(*courses.toTypedArray())
    }

    override fun deleteCourse(course: Course) {
        db.courseDao.delete(course)
    }

    override fun deleteAllOnlineCourse() {
        deleteCourse(getCourses(false))
    }

    override fun getSyllabusSetting(): SyllabusSetting {
        var setting = db.syllabusSettingDao.syllabusSetting(account)
        if (setting == null) {
            setting = SyllabusSetting()
            setting.account = account
            var js = false
            for (c in getAllCourses()) {
                if (c.address.contains("旗教")) {
                    js = true
                    break
                }
            }
            val times = ArrayList<Int>()
            Collections.addAll(times, *SyllabusSetting.intBeginTime[if (js) 1 else 0])
            setting.beginTime = times
        }
        return setting
    }

    override fun saveSyllabusSetting(syllabusSetting: SyllabusSetting) {
        db.syllabusSettingDao.save(syllabusSetting)
    }

    override fun getAllScores(): List<Score> {
        return db.scoreDao.allScores(account)
    }

    override fun getScoresByYear(year: String): List<Score> {
        return db.scoreDao.allScoresByYear(account, year)
    }

    override fun getScoresByTerm(term: String): List<Score> {
        return db.scoreDao.allScoresByTerm(account, term)
    }

    override fun getScoreById(id: Long): Score {
        return db.scoreDao.getScoreById(id)
    }

    override fun getScores(year: String, term: String): List<Score> {
        return db.scoreDao.allScores(account, year, term)
    }

    override fun deleteScore(year: String, term: String) {
        db.scoreDao.delete(year, term)
    }

    override fun deleteAllScore() {
        db.scoreDao.delete(account)
    }

    override fun saveScore(score: Score) {
        db.scoreDao.save(score)
    }

    override fun saveScore(scores: List<Score>) {
        db.scoreDao.save(*scores.toTypedArray())
    }

    override fun getAllExams(): List<Exam> {
        return db.examDao.allExams(account)
    }

    override fun getExams(year: String, term: String): List<Exam> {
        return db.examDao.allExams(account, year, term)
    }

    override fun saveExam(exams: List<Exam>) {
        db.examDao.save(*exams.toTypedArray())
    }

    override fun getToken(account: String): Token? {
        return db.tokenDao.token(account)
    }

    override fun saveToken(token: Token) {
        db.tokenDao.save(token)
    }

    override fun clearAllData() {
//        for (allDao in DaoManager.getInstance().daoSession.getAllDaos()) {
//            allDao.deleteAll()
//        }
    }

    override fun getYearTermList(): YearTerm {
        val yearList: MutableList<String> = ArrayList()
        val c = Calendar.getInstance()
        c.add(Calendar.MONTH, 8)
        val year = c[Calendar.YEAR]
        val acYear: Int //几几届学生
        val account = getInUseUser()?.account!!
        acYear = if (account.length == 10) {
            account.substring(1, 3).toInt() + 2000
        } else {
            account.substring(0, 2).toInt() + 2000
        }
        for (i in 0 until year - acYear) {
            yearList.add(String.format(Locale.CHINA, "%d-%d", year - i - 1, year - i))
        }
        val termList: MutableList<String> = ArrayList()
        for (i in 1..3) {
            termList.add(i.toString())
        }
        return YearTerm(yearList, termList)
    }

    override fun getYearTerm(): Pair<String, String> {
        val c = Calendar.getInstance()
        c.add(Calendar.MONTH, 6)
        val toTerm = if (c[Calendar.MONTH] < 8) "1" else "2"
        val year = c[Calendar.YEAR]
        val toYear = String.format("%d-%d", year - 1, year)
        return Pair(toYear, toTerm)
    }

    override fun getElecQuery(): ElecQuery? {
        return db.elecQueryDao.elecQuery(account)
    }

    override fun saveElecQuery(elecQuery: ElecQuery) {
        db.elecQueryDao.save(elecQuery)
    }

    override fun getElecCookie(): ElecCookie? {
        return db.elecCookieDao.elecCookie(account)
    }

    override fun saveElecCookie(cookie: ElecCookie) {
        db.elecCookieDao.save(cookie)
    }

    override fun getElecUser(): ElecUser? {
        return db.elecUserDao.elecUser(account)
    }

    override fun saveElecUser(elecUser: ElecUser) {
        db.elecUserDao.save(elecUser)
    }

    override fun getGlobalSetting(): GlobalSetting {
        var setting = db.globalSettingDao.globalSetting(account)
        if (setting == null) {
            setting = GlobalSetting()
            setting.account = account
            saveGlobalSetting(setting)
        }
        return setting
    }

    override fun saveGlobalSetting(setting: GlobalSetting) {
        db.globalSettingDao.save(setting)
    }

}